feat: Generic UDP forwarding with JSON parsing option

figuernd opened the pull request:
This makes the UDP-to-ROS node more generic so that it can also be used for capturing ship network data broadcasts and published to web node.

Also disables profiler if winch is not in use so that IFCB arm can be run without movement.

johnwaalsh committed:
feat: initial digital logger node implementation

johnwaalsh committed:
correct logger request format and add example config

johnwaalsh committed:
remove unused LoggerStatus message type

figuernd committed:
feat: Generic UDP forwarding with JSON parsing option

johnwaalsh committed:
update import order to follow pep8 convention

johnwaalsh committed:
switch from absolute rospy param paths to relative

johnwaalsh committed:
update requested outlet numbers to dynamically change with config file

johnwaalsh committed:
add parameter to enable/disable digital logger node from yaml

johnwaalsh committed:
add optional comment for digital logger node in yaml

johnwaalsh committed:
switch to list to track outlet publishers and simplify outlet status checking loop

johnwaalsh committed:
fix redundant string construction and remove auth from logging

johnwaalsh committed:
switch yaml to start at 0 instead of 1

johnwaalsh committed:
remove unused subscriber variable assignment

johnwaalsh committed:
update yaml to start at 0 instead of 1

johnwaalsh committed:
update to remove requests dependency

rgov commented:
Did we ever figure out why the DS ROS one doesn't work for you?

rgov commented:
What's your thought on using a class for this? I am in the minority but I wrote an argument against this in John's PR.

figuernd committed:
refactor: Moved udp_to_ros to use preferred classless node paradigm

figuernd commented:
> Did we ever figure out why the DS ROS one doesn't work for you?

No. IIRC nothing was showing up on the topic output and I wasn't getting any warnings or errors to indicate issues, and I ran out of time to do more trial and error. Not against trying again at some point, but this has been working for now.

> What's your thought on using a class for this? I am in the minority but I wrote an argument against this in John's PR.

Touche. Refactored into classless pattern.

rgov commented:
Thank you.

We should consider a few points about the JSON parsing. It works for only a very limited subset of valid JSON inputs:

* It expects that the JSON input is an object (dictionary), though it's valid to have a top-level list, string, or scalar.

        >>> json.loads('[1,2,3]')
        [1, 2, 3]

* It does not handle nested structures.
* It does not handle heterogeneous lists (`[1, true]`).
* It does not handle lists of strings, booleans, and other types that are relatively straightforward to map.
* It assumes all integers and floats can be represented by the types `Int32` and `Float32`, which isn't required by the JSON spec.
* The coercion of all other types to strings raises a question about encoding, which ROS is picky about.
* It assumes all object keys will lead to valid ROS topic names.

A higher level observation is that this conflates two operations, bridging UDP services with ROS, and breaking apart JSON content that is stored in a `RawData` message. If these were factored into two different nodes, the JSON parsing could be applied to any type of packet source (e.g., TCP or serial).

The last point to discuss is whether it's usually desirable for object keys to be exploded into separate topics, rather than parse to a complex data type. I think there's precedent to doing it this way, like AML messages, where it's comprised of separate channels that we publish separately. However it would be awkward to split `{"latitude": 36.056595, "longitude": -112.125092}` into separate topics.

johnwaalsh committed:
switch to urllib from http.client

johnwaalsh committed:
switch OutletStatus from string to bool

johnwaalsh committed:
switch to functional design

figuernd committed:
fix: More purposeful naming, better type checking

figuernd commented:
Good points - 


>     * It expects that the JSON input is an object (dictionary), though it's valid to have a top-level list, string, or scalar.

Yeah this is bad naming on my part - in reality we want a JSON-dict-to-subtopic parser for the express purpose of making JSON dictionaries published on ships easy to publish on the webnode. So this node enables someone to deploy PhytO-ARM to a ship and start publishing data to the webnode with just YAML configuration.

>     * It does not handle nested structures.

Yeah we really just want top level keys as subtopics. A long/lat object would be better as a string, as you note below, rather than recursively exploded.

>     * It does not handle heterogeneous lists (`[1, true]`).

Correct, unsure if there's a ROS-clean way to do so, so I just convert those to strings.

>     * It does not handle lists of strings, booleans, and other types that are relatively straightforward to map.

Added bool handling. Any other types you had in mind?

>     * It assumes all integers and floats can be represented by the types `Int32` and `Float32`, which isn't required by the JSON spec.

Right, I've increased to 64 to increase coverage but apart from making it a string I don't have a ROS-native way to handle an infinite number. Don't see an obvious BigNum type.

>     * The coercion of all other types to strings raises a question about encoding, which ROS is picky about.

Thoughts on how to improve this?

>     * It assumes all object keys will lead to valid ROS topic names.

Good point, I've added a check.

> A higher level observation is that this conflates two operations, bridging UDP services with ROS, and breaking apart JSON content that is stored in a `RawData` message. If these were factored into two different nodes, the JSON parsing could be applied to any type of packet source (e.g., TCP or serial).

True- could break the JSON exploding into a utility lib (I think that would be more appropriate than launching a separate node?), but I would rather save that refactor for when the need arises.
